### ISSUE TITLE
RocksDB arangorestore time degeneration on dead documents

### ISSUE THREAD

**Issue by tomterl** [arangodb/arangodb#6379]
## My Environment

* __ArangoDB Version__:        3.1.16 (dump), 3.4-DEV (restore)
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Manual Start
* __Configuration__:               
* __Infrastructure__:               own
* __Operating System__:        Ubuntu 16.04 
* __Used Package__:              arango-supplied Ubuntu .debs 

## Component, Query & Data

__Affected feature__:
restore is slowed down -- to a crawl in simple Scenarios, to a restore-failing timeout on bigger databases

__Dataset__:
Created with

````
FOR i in RANGE(1,200000)
   INSERT {'number': i} IN docs
````

__Replication Factor & Number of Shards (Cluster only)__:

## Steps to reproduce

1. Create a new database, e.g. `test_dead_docs_rocks_import` 
2. Add a collectionm `docs`
3.  Create 200.000 simple documents with
````
FOR i in RANGE(1,200000)
   INSERT {'number': i} IN docs
````
4. `arangodump` the database
5. `arangorestore` into a different instance using `rocksdb`
    This should take a couple of seconds (~5s in my case)

6. Create a new database, e.g. `test_dead_rocks_import`
7. Add a collection `docs`
8 . Create 200.000 simp;le documents with the query 
````
FOR i in RANGE(1,200000)
   INSERT {'number': i} IN docs
````
9. Truncate the docs-collection
10. Re-Create the 200.000 documents by executing the statement again
11. `arangodump` the database
12. `arangorestore` into a different instance using `rocksdb`
     This takes now insufferable long, although the same amount of data (the same documents, content-wise) is inserted into the db -- ~95s in my case
 
__Problem__:

If a database has seen a lot of changes (deletions/insertions), it's impossible to import it in an instance using rocksdb -- mmfiles doesn't show this problem.  This is the simplest case i could think of, to reproduce this problem.  We had this problem with our real-life production environment; we could work around it by creating a database totally fresh and transfer this data; but I fear in the future we will need to be able to transfer data with a lot of edit history without the ability to recreate the current state without that history.

__Expected result__:

deletion/reinsertion history of documents shouldn't affect restore-times.

**Comment by graetzer** [comment]
I just tried it on 3.4 and couldn't reproduce this behaviour, both dumps had roughly the same size and took about the same amount of time.
 
Maybe I did not understand this properly: 
Can you be more more specific in what is taking so long, just the last arangorestore step (after truncating + reinserting 200k docs) ? 

Is the size of the dump output directory's bigger when you truncated the collection ?

**Comment by graetzer** [comment]
Alright so you are dumping from _mmfiles_ into a _rocksdb_ instance ?
That would narrow it down a little.

**Comment by tomterl** [comment]
Yes - should have added that to the version info...

**Comment by jsteemann** [comment]
Just to rule out one further potential cause: *infrastructure* above reads *own*, so I assume the target arangod you are restoring into is not running on AWS and thus is not affected by running out of I/O burst credits. Can you confirm that it's not running on AWS and that I/O performance is expected to be stable over time in your environment?

**Comment by tomterl** [comment]
Not AWS - one virtual machine (big), one workstation class machine -- both runnnig Ubuntu,both showing the same behavior --both with plenty of resources to spare while the restore is running

**Comment by jsteemann** [comment]
Ok, thanks for confirming. Then it should be pretty reproducible for us.

**Comment by graetzer** [comment]
Hey I just tried witha devel version of 3.4 and could not reproduce the issue. You wrote that you used  3.4-DEV for restore; Was this a nightly version, or did you use the earlier Milestone release ?
Can you perhaps retry this with the latest RC, that we released last week ?

**Comment by tomterl** [comment]
The current nightly (just installed via apt-get from https://www.arangodb.com/repositories/nightly/xUbuntu_16.04) shows the same behaviour -- I'll try it with RC packages.

**Comment by tomterl** [comment]
3.4-RC1 shows the same behaviour -- although it took >110 seconds now (very possibly due to other load on the machine)

**Comment by vinaypyati** [comment]
Hi,

I tried reproducing this issue on Windows and following are the steps.
1. On version 3.1.16, create a database "test", collection "docs" and insert 200k documents. 
2. `arangodump` the "test" database - _dump1_
3. While being on the same version, truncate the "docs" collection and re-create 200k documents. Also, make any modifications to any of the documents.
4. Again, `arangodump` the "test" database - _dump2_

**Size**
Size of _dump1_: 30M
Size of _dump2_: 60M

Now,  `arangorestore` dumps using 3.4.0 RC1(RocksDB). 

Time taken to restore _dump1_: 12.3s
Time taken to restore _dump2_: 36.4s

`arangodump` preserves the revisions when there are subsequent dumps - meaning that if there are 3 modifications made to a document between dump1, dump2 and dump3, _dump3_ will contain all the 3 revisions. This seems to increase the size of the dump. @graetzer is this intended behaviour? 

More internal discussions regarding the revisioning is [here](https://github.com/arangodb/planning/issues/2799)

**Comment by tomterl** [comment]
Hmm - @vinaypyati : you see a factor of 3, I have almost a factor of 20 in time; while your dump increases by a factor of three (allowing for a factor of 3 in time) -- my dump is only 1.5 times the size.

Do you see the same stats when importing into an mmfiles instance? Can you test that?

My problem is not per-se with increasing time/memory needs in case of modifications; but with a stalling (on 92%, document 180185 in my case) during import into a rocksdb instance; mmfiles-backend works just fine.

**Comment by vinaypyati** [comment]
@tomterl : in my case, the dump size seems inconsistent too on further exports. 

Yes, I see same stats when importing into MMFiles instance. 

> I could just upload the offending export as zip/tgz somewhere? tgz is ~4.3M

I guess you can drag and drop the zip/gz in this issue. This will help us in reproducing the actual problem. 

Thanks.

**Comment by tomterl** [comment]
Here it is:
[docs_export_slow.tar.gz](https://github.com/arangodb/arangodb/files/2383092/docs_export_slow.tar.gz)

**Comment by jsteemann** [comment]
I just analyzed the issue with the provided dump with current 3.4.
The slowness is caused by the dump containing a lot of duplicate keys.
For example (with line numbers): 
```
2:{"type":2300,"data":{"_id":"docs/23346823561069","_key":"23346823561069","_rev":"_XYgkHAu--j","number":38}}
395722:{"type":2300,"data":{"_id":"docs/23346823561069","_key":"23346823561069","_rev":"_XYgrOT2-_X","number":38}}
```
or
```
9913:{"type":2300,"data":{"_id":"docs/23346823908717","_key":"23346823908717","_rev":"_XYgkIMS--i","number":173861}}
398574:{"type":2300,"data":{"_id":"docs/23346823908717","_key":"23346823908717","_rev":"_XYgrOU6-_B","number":173861}}
```
So many many duplicate keys keys. This is how data will come out of MMFiles when dumping a collection into which a document is inserted and then later updated. In the dump, there is no distinction between insert and update/replace, so this is expected.

However, when restoring a dump, it is unknown for each row in the dump whether the document already exists or not. The server will attempt to insert the document first, and if it does already exist, there will be a unique constraint violation, and a replace is tried. In the end, this will lead to the same data as in the dump file.

This requires that when the speculative insert fails with a unique constraint violation, the insert operation needs to be rolled back as if it never happened. This roll back is a very expensive operation in RocksDB, because we are processing the restore data in batches of several thousand items. Let's assume we already have 1000 operations inside a batch, and then an insert fails. The failing insert will not only pop the latest operation(s) from the batch but entirely rebuild it. 
So when there are already 1000 operations in a batch and operation 1001 fails, this will lead to rewriting the 1000 operations in the batch. So if there are some 100K documents that trigger unique constraint violations, this will quickly sum up and make the operations very expensive.

I will look into some ways to optimize this tomorrow. One obvious fix would be to check first if a target document with the specified key exists.

**Comment by jsteemann** [comment]
Should be addressed via https://github.com/arangodb/arangodb/pull/6574

**Comment by jsteemann** [comment]
The patch decreases the restore time for the provided dump from 60s to 5s.

**Comment by tomterl** [comment]
Cool! Will this be in nightly-packages? If so, I'll update and check the behavior first thing next week.

**Comment by jsteemann** [comment]
Merged back into 3.4 and devel branches. The bug fix will be part of 3.4RC2.

**Comment by jsteemann** [comment]
No worries. Thanks for your willingness to test it!

**Comment by sleto-it** [comment]
3.4-rc 2 is available for download - please let us know if we can consider this resolved

**Comment by tomterl** [comment]
Just retested: Looks good! No stalling, finishes in ~10s on my machine, which is expected. Thanks a lot.

**Comment by jsteemann** [comment]
For info, for the 3.3 branch there is now also a PR addressing this issue:
https://github.com/arangodb/arangodb/pull/7438
The fix is supposed to be released with 3.3.20.
